### ISSUE TITLE
Spring Cloud Data Flow does not work how is expected for file sink and curl client

### ISSUE THREAD

**Issue by manueljordan** [spring-cloud/spring-cloud-dataflow#1587]
In Mac, working with `Spring XD`  

* `1.3.2.RELEASE`

The following works fine:

`stream create --name orderFilter --definition "http --port=9090 | filter --expression=#jsonPath(payload,'$.countryCode').equals('US') | file --name=manolito.txt --dir=/Users/manueljordan/Desktop --mode=APPEND" --deploy`

Through its own `shell` the following works:

    http post --target http://localhost:9090 --data "{\"id\":\"2773\",\"countryCode\":\"US\",\"orderAmt\":\"100\"}"
    http post --target http://localhost:9090 --data "{\"id\":\"2774\",\"countryCode\":\"CA\",\"orderAmt\":\"125\"}"
    http post --target http://localhost:9090 --data "{\"id\":\"2775\",\"countryCode\":\"US\",\"orderAmt\":\"110\"}"

The data is filtered and the output is located in `/Users/manueljordan/Desktop` how is expected.

Until here all is ok.

Now migrating to `Spring Cloud Data Flow` working with

* `spring-cloud-dataflow-server-local-1.2.2.RELEASE.jar`
* `spring-cloud-dataflow-shell-1.2.2.RELEASE.jar`

Through the GUI (`http://localhost:9393/dashboard/index.html#/streams/definitions`) I can create the same stream. 

Again through the shell (from _spring-cloud-dataflow-shell_) the following works:

      http post --target http://localhost:9090 --data "{\"id\":\"2773\",\"countryCode\":\"US\",\"orderAmt\":\"100\"}"
      http post --target http://localhost:9090 --data "{\"id\":\"2774\",\"countryCode\":\"CA\",\"orderAmt\":\"125\"}"
      http post --target http://localhost:9090 --data "{\"id\":\"2775\",\"countryCode\":\"US\",\"orderAmt\":\"110\"}"

The same http post used for `Spring XD`

**Situation 01** even when the three `http post` from above work about the filter behaviour, the output file is **not** created in `/Users/manueljordan/Desktop`, it is created in `/private/var/folders/nf/ql69b5zd7w54jw6nb5zmvck80000gn/T/dataflow/output` instead

Observe the `dataflow` directory within `ql69b5zd7w54jw6nb5zmvck80000gn`

Why this behaviour? How fix this?, some special configuration missing?

**Situation 02** about *cURL*, according with the reference documentation, in the [10.1. Endpoints][2] section, just practically below the `Figure 3. The Spring Cloud Data Flow Server` shows that `cURL` can be used too.

Through `iterm 2` if I use the following commands: (it according with [How to POST JSON data with Curl from Terminal/Commandline to Test Spring REST?][1]) 

    curl -H "Content-Type: application/json" -X POST -d '"{\"id\":\"2773\",\"countryCode\":\"US\",\"orderAmt\":\"100\"}"' http://localhost:9090
    curl -H "Content-Type: application/json" -X POST -d '"{\"id\":\"2774\",\"countryCode\":\"CA\",\"orderAmt\":\"125\"}"' http://localhost:9090
    curl -H "Content-Type: application/json" -X POST -d '"{\"id\":\"2775\",\"countryCode\":\"US\",\"orderAmt\":\"110\"}"' http://localhost:9090

Nothing happens (I press enter to execute the command and just pass to the new line prompt, that's all). Why? what is missing?

Post question shared in stackoverflow:

[Spring Cloud Data Flow does not work how is expected for file sink and curl client](https://stackoverflow.com/questions/44933944/spring-cloud-data-flow-does-not-work-how-is-expected-for-file-sink-and-curl-clie)

  [1]: https://stackoverflow.com/questions/7172784/how-to-post-json-data-with-curl-from-terminal-commandline-to-test-spring-rest
  [2]: http://docs.spring.io/spring-cloud-dataflow/docs/1.2.2.RELEASE/reference/htmlsingle/#arch-data-flow-server-endpoints

**Comment by sabbyanandan** [comment]
Hi, @manueljordan.

For "Situation 01", you'd need to pass the property as `--directory` - please refer to the [docs](http://docs.spring.io/spring-cloud-stream-app-starters/docs/Bacon.RELEASE/reference/html/spring-cloud-stream-modules-sinks.html#_options_36). Going forward, for application specific issues, please find the respective repo from [spring-cloud-stream-app-starters](https://github.com/spring-cloud-stream-app-starters/) ORG instead of creating an issue in SCDF project. 

For "Situation 02", what do you see in the `http-source` logs? Can you try without all the quotes, simply: `{"id":"2775","countryCode":"US","orderAmt":"110"}`. As far as it is a valid JSON, it should be acceptable.

Lastly, it'd be great if you can avoid duplicate posts.

**Comment by trisberg** [comment]
Hi,

We have changed some of the options for the Spring Cloud Stream App Starters so they don't always match up with what we use for Spring XD. In your case the `--dir` option is now `--directory` -- see [File Sink docs](http://docs.spring.io/spring-cloud-stream-app-starters/docs/Bacon.RELEASE/reference/html/spring-cloud-stream-modules-sinks.html#spring-cloud-stream-modules-file-sink).

For the `curl` command you have escaped the quotes which isn't needed and probably interfere with your filter. Try using this:

```
curl -H "Content-Type: application/json" -X POST -d '{"id":"2773","countryCode":"US","orderAmt":"100"}' http://localhost:9090
curl -H "Content-Type: application/json" -X POST -d '{"id":"2774","countryCode":"CA","orderAmt":"125"}' http://localhost:9090
curl -H "Content-Type: application/json" -X POST -d '{"id":"2775","countryCode":"US","orderAmt":"110"}' http://localhost:9090
```

Let me know if this helps.

**Comment by manueljordan** [comment]
Even when `--directory` works, why is documented in [File Sink docs](http://docs.spring.io/spring-cloud-stream-app-starters/docs/Bacon.RELEASE/reference/html/spring-cloud-stream-modules-sinks.html#spring-cloud-stream-modules-file-sink) how `file.directory` instead?. It is confuse in some way.

**Comment by trisberg** [comment]
Thanks for giving an update and some feedback on the docs. I would agree that the docs need some improvements, especially around the getting started experience and the app properties. We have some work [in progress](https://github.com/spring-cloud/spring-cloud-dataflow/issues/1573) for the next release. 

Spring Cloud Data Flow is based on Sprig XD but it is a complete rewrite and some things have changed. Some modules were turned into boot apps and in th eprocess we modified some properties that were inconsistent across all apps. The `--dir` property was one of them that got renamed.

The apps you deploy are just plain Spring Boot apps, completely decoupled from Data Flow, while the modules in XD where tightly coupled. This means that we don't know all possible properties for the app so we can't throw an exception if you use a property that doesn't exist.

For the use of "short form" properties like `--port` vs long form `--server.port` there is some documentation [here](http://docs.spring.io/spring-cloud-dataflow/docs/current/reference/html/spring-cloud-dataflow-create-stream.html#_application_properties)

Regarding the defaults, the use of <none> is sometimes misleading since there is a de-facto default value if you leave it out. I created an [issue](https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/108) for this.

So, bottom line is we are working on improving the docs and value your feedback.

**Comment by sabbyanandan** [comment]
Thanks for the detailed response, @manueljordan. 

> what does ORG mean?

I was referring to GitHub Organization. 😃 

> Note: Each time I restart the server all the apps are gone. I am assuming it is the default behaviour. I will do a research about how persist that data in some way, in case of a blackout.

SCDF uses an [embedded H2 database](http://docs.spring.io/spring-cloud-dataflow/docs/1.2.2.RELEASE/reference/htmlsingle/#getting-started-system-requirements) by default. Upon the server restarts, the in-memory persistence is lost. To avoid that, you could use an [external RDBMS](http://docs.spring.io/spring-cloud-dataflow/docs/1.2.2.RELEASE/reference/htmlsingle/#configuration-rdbms). Also, the Local-server is intended to be used _only_ for development and experimentation. For production deployments, it is recommended that you switch to a [platform implementation](http://cloud.spring.io/spring-cloud-dataflow/#spring-cloud-data-flow-implementations) such as Cloud Foundry or Kubernetes.

As Thomas pointed out, there are still gaps in the application docs. The goal is to review all the application READMEs via spring-cloud-stream-app-starters/app-starters-release#99 and generally also the docs restructuring for the upcoming release. Thanks for your patience.

**Comment by manueljordan** [comment]
Hello @trisberg and @trisberg 

Thanks a lot by the information. Again the comments were written with the best intentions and how feedback.

I would be happy to contribute in some way, but I need your guidance first.

* Is there some Test approach to test a _module_ (source, processor, sink) scenario? Such as `http`? it specially about to validate an attribute, to throw an exception. And of course even when the attribute key is valid, validate the value, the following should be invalid `--server.port=abc`. From there we can improve the documentation quickly by itself and from the migration. I am assuming a special class test template exists available for each module to test their attributes.
* From above, how do that always automatically? in offline or online mode? It about running all the infrastructure, Apache Kafta, Zookeeper etc. Just curious how you handle that.
* Again about documentation, why many modules has a prefix about its attributes? such as `file.directory`. It about the `file` sink

**Comment by sabbyanandan** [comment]
> Is there some Test approach to test a module (source, processor, sink) scenario? Such as http? it specially about to validate an attribute, to throw an exception. And of course even when the attribute key is valid, validate the value, the following should be invalid --server.port=abc. From there we can improve the documentation quickly by itself and from the migration. I am assuming a special class test template exists available for each module to test their attributes.

> From above, how do that always automatically? in offline or online mode? It about running all the infrastructure, Apache Kafta, Zookeeper etc. Just curious how you handle that.

There are a few testing layers to this. [1] Unit-tests live within the App itself (e.g., see [file-sink](https://github.com/spring-cloud-stream-app-starters/file/blob/master/spring-cloud-starter-stream-sink-file/src/test/java/org/springframework/cloud/stream/app/file/sink/FileSinkTests.java)) [2] App level integration tests (it's lacking) [3] End-to-end acceptance plays a critical role in testing the App in SCDF (e.g., see [http-source](https://github.com/spring-cloud/spring-cloud-dataflow-acceptance-tests/blob/master/src/test/java/org/springframework/cloud/dataflow/acceptance/test/HttpSourceTests.java#L37)). We have CI plans to test them against incremental commits. 

What we are also working on is a [detailed write-up](https://github.com/markpollack/spring-cloud-dataflow-samples/blob/dev-marketing/docs/src/main/asciidoc/stream-apps.adoc) on how someone would go about defining a problem to design, test, and using it in SCDF (e.g., spring-cloud/spring-cloud-dataflow-samples#27 | spring-cloud/spring-cloud-dataflow-samples#28). Once we have the branch merged, it'd be a great area for collaboration with the community in improving the overall material. 

> Again about documentation, why many modules has a prefix about its attributes? such as file.directory. It about the file sink

The Apps can be used standalone. They do not have any dependency with SCDF. To make the App properties more descriptive, we have prefixes to make its properties easy to identify and as well to avoid collisions with regular Spring Boot properties. When using them in SCDF, however, we whitelist the properties to make it even more easier for someone to pick them up in DSL/Flo.

**Comment by manueljordan** [comment]
@sabbyanandan thanks a lot by the detailed information, seems I should await a little more to get a detailed guidance about instructions about how to get involved with these apps about documentation and testings. It according your three latest links shared.

Be sure I want participate in this project. But I need a detailed guidance _before_ to get involved.

Best!

**Comment by sabbyanandan** [comment]
Hi, @manueljordan! This has been in the backlog for a long time; we are sorry for not getting back to you soon. Based on the feedback here and from the community/customers, we happened to launch a standalone developed-focused [microsite](https://dataflow.spring.io/) for SCDF and the ecosystem — primarily, Spring Cloud Stream, Spring Cloud Task, and the Apps.

We are continuing to incrementally add more of end-to-end reference architectures, [recipes](https://dataflow.spring.io/docs/recipes/), and [developer guides](https://dataflow.spring.io/docs/stream-developer-guides/). Though we don't yet have an entirely focused sample for every stream/task app that we ship, it is an area that we can further improve with the community contributions from you and others. 

Please have a look and let us know if you have any other feedback on the Microsite. You can open new issues [here](https://github.com/spring-io/dataflow.spring.io/issues).

**Comment by manueljordan** [comment]
Thanks for the new links .... I will do a research and participation when be appropriate (time is tight)
